[16.0][MIG] stock_orderpoint_mto_as_mts: Migration to 16.0

QuocDuong1306 opened the pull request:
This PR takes over:
- https://github.com/OCA/stock-logistics-orderpoint/pull/14

Notes:

- Renamed `sale_stock_mto_as_mts_orderpoint` to `stock_orderpoint_mto_as_mts`
- Removed dependency on stock_orderpoint_manual_procurement as not needed anymore, see https://github.com/OCA/stock-logistics-warehouse/pull/1641
- Used `is_mto` on product (depend on product-attribute/product_route_mto)
- Used `is_mto` on route instead of xmlid (provided by stock-logistics-warehouse/stock_route_mto - already pulled by product_route_mto)
- Performed the bellow in `product.product`:
- Processed: the orderpoint must be created only for warehouses where there is no mto rule for that warehouse
- Moved the orderpoint archiving from product.template and refactored it
- Made the orderpoint archiving configurable on the warehouse
- Dropped `is_from_mto_route` on stock.move. That should not be in this module, not related to orderpoint

grindtildeath committed:
Create module sale_stock_mto_as_mts_orderpoint

grindtildeath committed:
Check move state in selection of delivery move

oca-travis committed:
[UPD] Update sale_stock_mto_as_mts_orderpoint.pot

OCA-git-bot committed:
[ADD] icon.png

OCA-git-bot committed:
sale_stock_mto_as_mts_orderpoint 13.0.1.2.1

OCA-git-bot committed:
[UPD] README.rst

sebalix committed:
[FIX] sale_stock_mto_as_mts_orderpoint: support product set

TDu committed:
[13.0][FIX] sale_stock_mto_as_mts_orderpoint

For user without proper rights the confirmation of a sale.order can lead
to an access error on the orderpoint creation.
The sudo fixes that.

OCA-git-bot committed:
sale_stock_mto_as_mts_orderpoint 13.0.1.2.2

dzungtran89 committed:
[IMP] sale_stock_mto_as_mts_orderpoint: black, isort, prettier

dzungtran89 committed:
[MIG] sale_stock_mto_as_mts_orderpoint: Migration to 14.0

sebalix committed:
[IMP] sale_stock_mto_as_mts_orderpoint: test coverage

oca-travis committed:
[UPD] Update sale_stock_mto_as_mts_orderpoint.pot

OCA-git-bot committed:
sale_stock_mto_as_mts_orderpoint 14.0.1.0.1

OCA-git-bot committed:
[UPD] README.rst

sergiocorato committed:
[FIX] black, isort, prettier

OCA-git-bot committed:
[BOT] post-merge updates

mymage committed:
Added translation using Weblate (Italian)

tuantrantg committed:
[IMP] sale_stock_mto_as_mts_orderpoint: apply pre-commit auto fixes

tuantrantg committed:
[REF] stock_orderpoint_mto_as_mts: Rename module from sale_stock_mto_as_mts_orderpoint

tuantrantg committed:
[IMP] stock_orderpoint_mto_as_mts

QuocDuong1306 commented:
Hi @jbaudoux, I applied your suggestions in https://github.com/OCA/stock-logistics-orderpoint/pull/14 to the last commit
BTW, added some tests also

jbaudoux commented on stock_orderpoint_mto_as_mts/models/product_product.py:
```diff
@@ -0,0 +1,131 @@
+            # No changes about is_mto
+            self.sudo()._ensure_default_orderpoint_for_mto()
+            return
+        mto_products = self._filter_mto_products()
+        # Find all MTO products that have been changed to Not MTO
+        products_to_update = original_mto_products - mto_products
+        if products_to_update:
+            products_to_update._archive_orderpoints_on_mto_removal()
```
This is only necessary when is_mto is removed
```suggestion
        if not vals.get("is_mto", True):
            # when is_mto is removed
            mto_products = self._filter_mto_products()
            # Find all MTO products that have been changed to Not MTO
            products_to_update = original_mto_products - mto_products
            if products_to_update:
                products_to_update._archive_orderpoints_on_mto_removal()
```

QuocDuong1306 replied:
Hello @jbaudoux , as I knew and tested, `is_mto` is the computed field, so it didn't appear in `vals` of `def write()`
So `not vals.get("is_mto", True)` is not worked in this case, that is the reason why I used `route_ids` instead of `is_mto`

jbaudoux replied:
It is defined on product.template and it appears on write vals of product.template but not on product.product as it is not defined there. That's why it is useless to hook on product.product.

QuocDuong1306 replied:
Yes, I know it, [It is defined on product.template](https://github.com/OCA/product-attribute/blob/3a646a09de5e0bc90f3a59328396b2a3da9d4771/product_route_mto/models/product_template.py#L11), but it is the computed field (and readonly = True in this case), if we want to change it, we must write on the dependency field (`route_ids`, `categ_id` and `is_mto of the route` in this case)

After that, `is_mto` on product.template will appear in vals of `def _write()`, not in `def write`

jbaudoux commented on stock_orderpoint_mto_as_mts/models/product_product.py:
```diff
@@ -0,0 +1,131 @@
+        return
+
+    def _filter_mto_products(self):
+        return self.filtered(lambda p: p.is_mto)
```
This method is useless. There is already the `is_mto` compute method

jbaudoux commented on stock_orderpoint_mto_as_mts/models/product_product.py:
```diff
@@ -0,0 +1,131 @@
+            self.sudo()._ensure_default_orderpoint_for_mto()
+            return
+        elif "route_ids" not in vals:
+            # No changes about is_mto
```
```suggestion
        if vals.get("is_mto"):
            # when is_mto is set
```

jbaudoux commented on stock_orderpoint_mto_as_mts/models/product_product.py:
```diff
@@ -0,0 +1,131 @@
+        self._update_mto_products(original_mto_products, vals)
+        return res
+
+    def _update_mto_products(self, original_mto_products, vals):
```
I would move this to product.template

jbaudoux commented on stock_orderpoint_mto_as_mts/models/product_product.py:
```diff
@@ -0,0 +1,131 @@
+        original_mto_products = self._filter_mto_products()
+        res = super().write(vals)
+        self._update_mto_products(original_mto_products, vals)
+        return res
```
Can be dropped, `is_mto` field is on template

jbaudoux commented on stock_orderpoint_mto_as_mts/models/product_template.py:
```diff
@@ -0,0 +1,20 @@
+    _inherit = "product.template"
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        template = super().create(vals_list)
+        template.product_variant_ids.sudo()._ensure_default_orderpoint_for_mto()
+        return template
```
Can be dropped. Orderpoint is on product

QuocDuong1306 replied:
If we drop this, will fail on case: Create product.template --> Create orderpoints

Because when create template --> template will trigger [here](https://github.com/odoo/odoo/blob/e3033d7aae5537403a1aedbae6e598514c04f703/addons/product/models/product_template.py#L433) --> for now, products will be created but `is_mto` is not computed (debugged) so we can not handle this case through `def create() of product.product`

jbaudoux reviewed (changes requested):
As we rely on module `product_route_mto` we should not re-implement the concept of `is_mto`. This is currently defined on product.template. In my opinion, there is extraneous complexity to cover the case where only a variant would be mto. It cannot be so far.

QuocDuong1306 commented:
Hi @jbaudoux , I updated the PR, and added more logics to handle cases that impact to `is_mto` on product (`stock.route`, `product.category`)

Based on [here](https://github.com/OCA/product-attribute/blob/3a646a09de5e0bc90f3a59328396b2a3da9d4771/product_route_mto/models/product_template.py#L16)

jbaudoux commented:
cc @rousseldenis

cyrilmanuel commented:
Hi @rousseldenis  is it good for you ?

jbaudoux commented on stock_orderpoint_mto_as_mts/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,13 @@
+This module aims to materialize the triggering of another stock rule through
+an orderpoint in the procurement of MTO products from a sale order.
```
```suggestion
an orderpoint in the procurement of MTO products.
```

jbaudoux commented on stock_orderpoint_mto_as_mts/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,13 @@
+using module `stock_orderpoint_manual_procurement`.
+
+Finally, orderpoints with min/max quantities of zero will be archived if the
+MTO route is removed on the product.
```
```suggestion
The MTO route will loose it's MTO rule as this is replaced by the generated orderpoints.

When a product is configured as MTO, then an orderpoint with min/max of zero is created. When the product is not anymore configured as MTO, the orderpoint can be automatically archived depending on the configuration you choose on the warehouse.
```

jbaudoux commented on stock_orderpoint_mto_as_mts/readme/CONFIGURATION.rst:
```diff
@@ -0,0 +1,12 @@
+* Change the Supply Method on stock.rule to Take From Stock, will ensure that
+  the need is materialized by the procure_recommended_qty on the orderpoint,
+  if the ensuing procurement (purchase order/manufacturing order) having the
+  orderpoint as origin is canceled.
```
```suggestion
In order to modify the MTO route and rely on generated orderpoints for MTO products instead of the MTO rule, you need to enable "mto as mts" on the warehouse
```

jbaudoux reviewed (commented):
Can you update the documentation with the performed modifications?

TDu commented on stock_orderpoint_mto_as_mts/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+{
+    "name": "Sale Stock Mto As Mts Orderpoint",
```
```suggestion
    "name": "Stock Orderpoint Mto As Mts",
```

QuocDuong1306 commented:
Hi guys, the PR is updated

jbaudoux commented:
/ocabot migration sale_stock_mto_as_mts_orderpoint

jbaudoux commented:
/ocabot merge nobump
